README.md
# vh5

> A Vue.js project

## Build Setup

``` bash
# install dependencies
npm install
npm i -g cross-env
npm i -g karma
```

# install JDK

从[http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html)下载JDK进行安装。

安装成功后运行`java -version`，看安装是否成功。

如果提示命令出错，则有可能是因为环境变量没有配置，先进行环境变量配置。

# serve with hot reload at localhost:8080

`npm run dev` 或者 `sudo npm run dev`

# build for production with minification
`npm run build` 或者 `sudo npm run build`

# build for production and view the bundle analyzer report
`npm run build --report` 或者 `sudo npm run build --report`

# run unit tests
`npm run unit` 或者 `sudo npm run unit`

# run e2e tests
npm run e2e

# run all tests
npm test
```

For detailed explanation on how things work, checkout the [guide](http://vuejs-templates.github.io/webpack/) and [docs for vue-loader](http://vuejs.github.io/vue-loader).

# original project structure

```
.
├── build/                      # webpack config files
│   └── ...
├── config/
│   ├── index.js                # main project config
│   └── ...
├── src/
│   ├── main.js                 # app entry file
│   ├── App.vue                 # main app component
│   ├── components/             # ui components
│   │   └── ...
│   └── assets/                 # module assets (processed by webpack)
│       └── ...
├── static/                     # pure static assets (directly copied)
├── test/
│   └── unit/                   # unit tests
│   │   ├── specs/              # test spec files
│   │   ├── index.js            # test build entry file
│   │   └── karma.conf.js       # test runner config file
│   └── e2e/                    # e2e tests
│   │   ├── specs/              # test spec files
│   │   ├── custom-assertions/  # custom assertions for e2e tests
│   │   ├── runner.js           # test runner script
│   │   └── nightwatch.conf.js  # test runner config file
├── .babelrc                    # babel config
├── .postcssrc.js               # postcss config
├── .eslintrc.js                # eslint config
├── .editorconfig               # editor config
├── index.html                  # index.html template
└── package.json                # build scripts and dependencies
```

### 生成后的vue-loader的options.json

```
{
  "loaders": {
    "css": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      }
    ],
    "postcss": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      }
    ],
    "less": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      },
      {
        "loader": "less-loader",
        "options": {
          "sourceMap": false
        }
      }
    ],
    "sass": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      },
      {
        "loader": "sass-loader",
        "options": {
          "indentedSyntax": true,
          "sourceMap": false
        }
      }
    ],
    "scss": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      },
      {
        "loader": "sass-loader",
        "options": {
          "sourceMap": false
        }
      }
    ],
    "stylus": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      },
      {
        "loader": "stylus-loader",
        "options": {
          "sourceMap": false
        }
      }
    ],
    "styl": [
      "vue-style-loader",
      {
        "loader": "css-loader",
        "options": {
          "minimize": false,
          "sourceMap": false
        }
      },
      {
        "loader": "stylus-loader",
        "options": {
          "sourceMap": false
        }
      }
    ]
  }
}
```

### 访问页面的地址

```
sudo npm run dev
```

然后地址栏输入:

http://localhost/home.html

配合 connect-history-api-fallback 的配置 以及 vue-router的配置之后，可以直接用：

http://localhost 访问 http://localhost/home.html
http://localhost/detail 访问 http://localhost/detail.html

一、目录结构

.happypack happypack打包是生成的目录
build  存放各种webpack打包的文件
config 主要是导出了一系列配置项，最主要的出口是config/index.js
dev    老项目的源码
dll    公共库打包过程中的文件
oldstatic   老项目通过老架构打包出来的代码，迁移到这里，目前新架构会暂时兼容它，把它拷贝到static对应的目录中
src    新架构源码
static 新架构生成的文件
test   测试目录


二、公共库的配置和打包：

在/config/common-lib-dependencies.js中修改公共库列表，这些公共库将通过dll来单独打包:

开发环境对应的dll打包命令是

```
npm run dll-dev
```

生产环境对应的dll打包命令是

```
npm run dll-build
```

配置完公共库之后，需要执行一次上述dll打包命令。

dll打包命令会将公共库打包到/dll/目录下，其中｀npm run dll-dev｀会打包到/dll/dev/下，而｀npm run dll-build｀会打包到/dll/prod/下，分别对应的是开发环境和生产环境。

虽然上述命令已经集成了将修改后的公共库配置打入项目各个页面的功能（详见package.json中，其中有 && sudo npm run dev 、 && sudo npm run build就是），但是需要注意的是/dll/目录只是一个过度性的目录，并非最终线上引用公共库打包文件的目录。


三、nginx的配置

配置两个地址，一个指向html，另一个指向附属的静态资源

server {
    listen       80;
    server_name  www.test.com;

    location / {
        root   /Users/dennis/work/newh5/static/html/;
        index  index.html;
    }

    location /wallet {
        root   /Users/dennis/work/newh5/static/html/;
        try_files $uri $uri/ /wallet.html;
        index  wallet.html;
    }

    location /help {
        root   /Users/dennis/work/newh5/static/html/;
        try_files $uri $uri/ /help.html;
        index  help.html;
    }

    location /testweb3 {
        root   /Users/dennis/work/newh5/static/html/;
        try_files $uri $uri/ /testweb3.html;
        index  testweb3.html;
    }
}

server {
    listen       80;
    server_name  static.test.com;


    location / {
        add_header 'Access-Control-Allow-Methods' 'GET,OPTIONS,PUT,DELETE' always;
        add_header 'Access-Control-Allow-Credentials' 'true' always;
        add_header 'Access-Control-Allow-Origin' '$http_origin' always;
        add_header 'Access-Control-Allow-Headers'  'Authorization,DNT,User-Agent,Keep-Alive,Content-Type,accept,origin,X-Requested-With,Qsc-Token,platform' always;
        if ($request_method = OPTIONS) {
            return 200;
        }

        root   /Users/dennis/work/newh5/static/;
        index  index.html;
    }
}


```
npm start
anywhere -p 9090
```



testchain/history
miner.start(1
)
miner.start(1)
web3.eth.personal
web3.personal
web3.personal.importRawKey("0xc6d035eabe029996b58ae42a09ad667a07bc01767ddc635c9a32e2723f689196")
web3.personal.importRawKey("0xc6d035eabe029996b58ae42a09ad667a07bc01767ddc635c9a32e2723f689196", "")
web3.personal.importRawKey({"version":3,"id":"19ba64ee-cabb-4e0f-b08b-45cee5cf8052","address":"5bdfea38428cdfda7470c7fbd77007125f598d7d","crypto":{"ciphertext":"5e4c38c21bf6ea233f2fc2ff7449668ba7cedfa308f2df37a0014055b2ba7259","cipherparams":{"iv":"9baf5d8946c2ad012bc57d816ea8b231"},"cipher":"aes-128-ctr","kdf":"scrypt","kdfparams":{"dklen":32,"salt":"a67942bba8e9b23213384ba47129c05ae923040e67b95fd0c6957eb05bda218b","n":8192,"r":8,"p":1},"mac":"2b0f4045ad7db29b04407fbb656c36d4ef256cf76ab66bc4e12c223b77442e13"}}, "")
web3.personal.importRawKey([{"version":3,"id":"19ba64ee-cabb-4e0f-b08b-45cee5cf8052","address":"5bdfea38428cdfda7470c7fbd77007125f598d7d","crypto":{"ciphertext":"5e4c38c21bf6ea233f2fc2ff7449668ba7cedfa308f2df37a0014055b2ba7259","cipherparams":{"iv":"9baf5d8946c2ad012bc57d816ea8b231"},"cipher":"aes-128-ctr","kdf":"scrypt","kdfparams":{"dklen":32,"salt":"a67942bba8e9b23213384ba47129c05ae923040e67b95fd0c6957eb05bda218b","n":8192,"r":8,"p":1},"mac":"2b0f4045ad7db29b04407fbb656c36d4ef256cf76ab66bc4e12c223b77442e13"}}], "")
web3.personal.importRawKey("0xc6d035eabe029996b58ae42a09ad667a07bc01767ddc635c9a32e2723f689196", "a")
web3.personal.importRawKey(0xc6d035eabe029996b58ae42a09ad667a07bc01767ddc635c9a32e2723f689196, "a")
web3.personal.importRawKey("0xc6d035eabe029996b58ae42a09ad667a07bc01767ddc635c9a32e2723f689196", "qwerty")
web3.personal
web3.personal.getListWallets
web3.personal.getListWallets(console.log)
web3.personal.getListAccounts(console.log)
web3.eth.coinbase
miner.start()
miner.stop()
miner.start(1)
miner
quit
eth.sendTransaction({to: "0x1F266c0A3d26447595b99DeD50b5adD0b1AC749D", value: 10000000000000000000})
eth.sendTransaction({to: "0x1F266c0A3d26447595b99DeD50b5adD0b1AC749D", value: 10000000000000000000, from: eth.coinbase})
quit
eth.sendTransaction({to: "0x1F266c0A3d26447595b99DeD50b5adD0b1AC749D", value: 10000000000000000000, from: eth.coinbase})
miner.start(1)
miner.stop()
eth.getBalance
eth.getBalance("0x1f266c0a3d26447595b99ded50b5add0b1ac749d")
web3.fromWei(eth.getBalance("0x1f266c0a3d26447595b99ded50b5add0b1ac749d"), "ether")
web3.fromWei(eth.getBalance(eth.coinbase), "ether")
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.start(10)
miner.start(3)
miner.stop()
eth.sendTransaction({to: "0x0212010e3F64a56FD6d7E33Db01279055B553BED", value: web3.toWei(100, "ether"), from: eth.coinbase})
miner.start()
miner.start(1)
miner.start(3)
miner.stop()
miner.start(3)
miner.stop()
miner.start(10)
miner.stop()
quit
eth.sendTransaction({from: eth.coinbase, to: "0xfBbBbCAD1dAE2665fB6Cdf0F0a18e86E2AC77285", value: web3.toWei(10, "ether")})
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
eth.getTransactionReceipt("0xa4c38a6a3f6ccc62a5a40037c9180f41ce228b918ec859fcc029d9f4208d0f78")
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
miner.stop()
miner.start(1)
0x434B4D25ddf654718a89c2B42690447C0Bf724AF
quit
miner.start(1)
miner.start(10)
miner.start(2)
miner.start()
miner.stop()
eth.get
eth
eth.getCode("0xb8060a6a964abec5bb0f8d9d379f28e5be1344f6")
eth.getTransactionReceipt("0x90eabf193e0bc6c429bb38f868e11948b0048875460fd6d04a1bf80a34a24e49")
eth.getTransaction("0x90eabf193e0bc6c429bb38f868e11948b0048875460fd6d04a1bf80a34a24e49")
miner.start()
miner.stop()
miner.start()
miner.stop
miner.stop()
miner.start()
miner.stop()
miner.start()
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.start(1)
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop
miner.stop()
miner.stop
miner.start(2)
miner.stop
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.stop()
miner.start(2)
miner.start(10)
miner.stop()
miner.start(10)
miner.stop()
eth.accounts
eth.getBalance(eth.accounts[0])

config/index.js
// see http://vuejs-templates.github.io/webpack for documentation.
var path = require('path');
var glob = require('glob');

var exportObj;

function joinPath(currentDir, subDir) {
  if(arguments.length > 1){
    return path.join(currentDir, subDir);
  } else {
    return path.join(__dirname, '..', currentDir);
  }
}

exportObj = {
  dev: {
    env: require('./dev.env'),
    port: 80,
    autoOpenBrowser: false,
    assetsSubDirectory: 'static',
    assetsPublicPath: '/',
    proxyTable: {
      // todo 待配置
      '/dll/lib.js': {
        target: 'http://127.0.0.1:9090',
        changeOrigin: true,
        pathRewrite: {
          '^/dll/': 'dll/dev/'
        }
      }
      // ,
      // '/api': {
      //   target: 'http://localhost:8545/',
      //   changeOrigin: true,
      //   pathRewrite: {
      //     '^/api': ''
      //   }
      // }
    },
    // CSS Sourcemaps off by default because relative paths are "buggy"
    // with this option, according to the CSS-Loader README
    // (https://github.com/webpack/css-loader#sourcemaps)
    // In our experience, they generally work as expected,
    // just be aware of this issue when enabling this option.
    cssSourceMap: false,                                     // css source map
    dll:{
      basePath: path.resolve(__dirname, '../dll/dev'),       //执行NODE_ENV=development webpack --config  build/webpack.dll.conf.js --progress时的输出目录
      fileName: 'lib.js',
      manifestFileName: 'manifest.json',
      outputPath: '/dll/',  // 生成目录，是相对于assetsSubDirectory的目录
      publicPath: '../../dll/'   // 注入地址，是相对于assetsSubDirectory的目录
    }
  },
  build: {
    env: require('./prod.env'),                             // NODE_ENV
    index: path.resolve(__dirname, '../static/html/home.html'),   // index page
    assetsRoot: path.resolve(__dirname, '../static'),         // assets root path
    assetsSubDirectory: './',                           // assets directory
    assetsPublicPath: '//static.test.com',                                  // assets public path
    productionSourceMap: false,                              // 是否生成source map
    // Gzip off by default as many popular static hosts such as
    // Surge or Netlify already gzip all static assets for you.
    // Before setting to `true`, make sure to:
    // npm install --save-dev compression-webpack-plugin
    compressor: {
      warnings: false,
      drop_console: true,
      drop_debugger: true
    },
    productionGzip: false,                                  // 是否开启Gzip
    productionGzipExtensions: ['js', 'css'],                // Gzip后缀名
    // Run the build command with an extra argument to
    // View the bundle analyzer report after build finishes:
    // `npm run build --report`
    // Set to `true` or `false` to always turn it on or off
    bundleAnalyzerReport: process.env.npm_config_report,     // ??process.env.npm_config_report获取的是--report参数吗

    dll:{
      basePath: path.resolve(__dirname, '../dll/prod'),       //执行NODE_ENV=production webpack --config  build/webpack.dll.conf.js --progress时的输出目录
      fileName: 'lib.js',
      manifestFileName: 'manifest.json',
      outputPath: '/js/dll/',  // 生成目录，是相对于assetsSubDirectory的目录
      publicPath: '//static.test.com/js/dll/'   // 注入地址，是相对于assetsSubDirectory的目录
    }
  },
  // 公共库的依赖配置
  commonLibDependencies: require('./common-lib-dependencies'),
  dllAlias: {
    // todo 这个地方待配置
    'vue$': 'vue/dist/vue.esm.js',
    '@': joinPath('src'),
    '@busi': joinPath('src/businessComponents'),
    '@com': joinPath('src/commonComponents')
  },
  alias: {
    pages: joinPath('src/pages'),
    utils:  'src/utils',
    filters: 'src/filters',
    directives:  'src/directives'
  },
  postcssConfigPath: path.resolve(__dirname, './postcss.config.js')
  // ,
  // pageEntries: {
  //   app: './src/main.js'
  // }
}

var globArr = [
  {
    name: 'page',
    filename: 'main'
  },
  {
    name: 'util',
    filename: 'util'
  },
  {
    name: 'filter',
    filename: 'filter'
  },
  {
    name: 'directive',
    filename: 'directive'
  }
];

var filePath = [];
var globInstance = [];

globArr.forEach((item, index) => {
  filePath[item.name] = exportObj.alias[item.name + 's'];
  globInstance[item.name] = new glob.Glob('**/' + item.filename + '.js', {
    cwd: filePath[item.name],
    sync: true  // 这里不能异步，只能同步
  });

  exportObj[item.name + 'Entries'] = {};

  globInstance[item.name].found.forEach((newItem) => {
    // console.log('>>>>>>>>>>\n 每一种自动生成的内容的类别：', item.name, filePath[item.name])
    exportObj[item.name + 'Entries'][newItem.replace(new RegExp('\/' + item.filename + '\.js'), '')] = joinPath(filePath[item.name], newItem);
  });
});

module.exports = exportObj;
